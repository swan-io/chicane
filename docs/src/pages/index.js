import Link from "@docusaurus/Link";
import useDocusaurusContext from "@docusaurus/useDocusaurusContext";
import CodeBlock from "@theme/CodeBlock";
import Layout from "@theme/Layout";
import * as React from "react";
import HomepageFeatures from "../components/HomepageFeatures";
import styles from "./index.module.css";

function HomepageHeader() {
  const { siteConfig } = useDocusaurusContext();
  return (
    <header className={"hero hero--primary " + styles.hero}>
      <img src={"./img/logo.svg"} alt="" className={styles.logo} />
      <div className="container">
        <h1 className={styles.heroTitle}>{siteConfig.title}</h1>
        <p className={styles.heroSubtitle}>{siteConfig.tagline}</p>
        <div>
          <Link className="button button--lg" to="/getting-started">
            Get started
          </Link>
          <span className={styles.separator} />
          <Link className="button button--lg" to="/top-level-api">
            API reference
          </Link>
        </div>
      </div>
      <div className={styles.code}>
        <CodeBlock className={styles.codeBlock} language={"typescript"}>
          {`import { createRouter } from "@swan-io/chicane";
import { match } from "ts-pattern";

export const Router = createRouter({
  Home: "/",
  UserList: "/users",
  UserDetail: "/users/:userId",
})

const App = () => {
  const route = Router.useRoute(["UserList", "UserDetail"]);

  return match(route)
    .with({name: "UserList"}, () => <UserList />)
    .with({name: "UserDetail"}, ({params: {userId}}) =>
      <UserDetail userId={userId} />
    )
    .otherwise(() => <NotFound />)
}
`}
        </CodeBlock>
      </div>
    </header>
  );
}

const Home = () => {
  const { siteConfig } = useDocusaurusContext();

  return (
    <Layout
      title={`Chicane: ${siteConfig.tagline}`}
      description="A safe router for React and TypeScript"
    >
      <HomepageHeader />
      <main>
        <HomepageFeatures />
      </main>
    </Layout>
  );
};

export default Home;
