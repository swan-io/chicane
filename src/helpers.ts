export const first = <T>(value: T[]): T | undefined => value[0];
export const isNonEmpty = (value: string): boolean => value !== "";
export const isParam = (value: string): boolean => value.startsWith(":");

export const isMultipleParam = (value: string): boolean =>
  value.startsWith(":") && value.endsWith("[]");

export const prefixWithSlash = (value: string): string =>
  value[0] === "/" ? value : `/${value}`;

export const concatPaths = (pathA: string, pathB: string): string => {
  const prefixedPathA = prefixWithSlash(pathA);
  const prefixedPathB = prefixWithSlash(pathB);

  return prefixedPathA === "/"
    ? prefixedPathB
    : prefixedPathB === "/"
    ? prefixedPathA
    : prefixedPathA + prefixedPathB;
};
