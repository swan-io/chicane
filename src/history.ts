/**
 * This module makes the different routes created with react-chicane listen to the same history instance
 */
import { createBrowserHistory, createPath } from "history";
import { useSyncExternalStore } from "use-sync-external-store/shim";
import { areLocationsEqual, decodeLocation } from "./location";
import { Location, Subscription } from "./types";

const subscriptions = new Set<Subscription>();

export const history = createBrowserHistory();

let currentLocation = decodeLocation(history.location, true);
let initialLocationHasChanged = false;

if (currentLocation.url !== createPath(history.location)) {
  history.replace(currentLocation.url); // URL cleanup
}

history.listen(({ location }) => {
  const nextLocation = decodeLocation(location, false);

  if (!areLocationsEqual(nextLocation, currentLocation)) {
    initialLocationHasChanged = true;
    currentLocation = nextLocation;
    subscriptions.forEach((subscription) => subscription(currentLocation));
  }
});

export const subscribe = (subscription: Subscription): (() => void) => {
  subscriptions.add(subscription);

  return () => {
    subscriptions.delete(subscription);
  };
};

export const getCurrentLocation = () => currentLocation;
export const hasInitialLocationChanged = () => initialLocationHasChanged;

export const useLocation = (): Location =>
  useSyncExternalStore(subscribe, getCurrentLocation);

// For testing purposes
export const resetInitialHasLocationChanged = () => {
  initialLocationHasChanged = false;
};
