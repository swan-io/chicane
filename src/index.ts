import * as React from "react";
import { useSubscription } from "use-subscription";
import { first } from "./helpers";
import {
  createPath,
  getCurrentLocation,
  hasLocationChanged,
  history,
  parsePath,
  subscribe,
  useLocation,
} from "./history";
import { getMatcher, match, matchAll, matchToHistoryPath } from "./matcher";
import {
  ExtractRoutesParams,
  GetNestedRoutes,
  Matcher,
  ParamsArg,
  PrependBasePath,
  Simplify,
} from "./types";

export { decodeSearch, encodeSearch } from "./search";
export type { Location, Search } from "./types";

const focusableElements: Record<string, boolean> = {
  A: true,
  INPUT: true,
  SELECT: true,
  TEXTAREA: true,
};

export const createRouter = <
  Routes extends Record<string, string>,
  BasePath extends string = string,
>(
  routes: Readonly<Routes>,
  options: {
    basePath?: BasePath;
    blockerMessage?: string;
  } = {},
) => {
  type RoutesWithBasePath = PrependBasePath<Routes, BasePath>;
  type NestedRoutes = GetNestedRoutes<RoutesWithBasePath>;
  type NestedRoutesParams = ExtractRoutesParams<NestedRoutes>;
  type FiniteRoutes = Omit<RoutesWithBasePath, keyof NestedRoutes>;
  type FiniteRoutesParams = ExtractRoutesParams<FiniteRoutes>;
  type RoutesParams = NestedRoutesParams & FiniteRoutesParams;

  const { basePath = "", blockerMessage = "" } = options;

  const matchers = {} as Record<keyof Routes, Matcher>;
  const rankedMatchers: Matcher[] = []; // higher to lower

  for (const routeName in routes) {
    if (Object.prototype.hasOwnProperty.call(routes, routeName)) {
      const matcher = getMatcher(routeName, `${basePath}/${routes[routeName]}`);
      matchers[routeName] = matcher;
      rankedMatchers.push(matcher);
    }
  }

  rankedMatchers.sort(
    (matcherA, matcherB) => matcherB.ranking - matcherA.ranking,
  );

  const goForward = (): void => history.forward();
  const goBack = (): void => history.back();

  const unsafeNavigate = (url: string): void => {
    const { pathname = "/", search = "", hash = "" } = parsePath(url);
    history.push({ pathname, search, hash });
  };

  const unsafeReplace = (url: string) => {
    const { pathname = "/", search = "", hash = "" } = parsePath(url);
    history.replace({ pathname, search, hash });
  };

  const createURL = <RouteName extends keyof FiniteRoutes>(
    routeName: RouteName,
    ...args: ParamsArg<FiniteRoutesParams[RouteName]>
  ): string => createPath(matchToHistoryPath(matchers[routeName], first(args)));

  const navigate = <RouteName extends keyof FiniteRoutes>(
    routeName: RouteName,
    ...args: ParamsArg<FiniteRoutesParams[RouteName]>
  ): void => history.push(matchToHistoryPath(matchers[routeName], first(args)));

  const replace = <RouteName extends keyof FiniteRoutes>(
    routeName: RouteName,
    ...args: ParamsArg<FiniteRoutesParams[RouteName]>
  ): void =>
    history.replace(matchToHistoryPath(matchers[routeName], first(args)));

  const useRoutes = <RouteName extends keyof FiniteRoutes | keyof NestedRoutes>(
    routeNames: ReadonlyArray<RouteName>,
  ): RouteName extends string
    ? { name: RouteName; params: Simplify<RoutesParams[RouteName]> }[]
    : never => {
    const routes = useSubscription(
      React.useMemo(() => {
        const matchers = rankedMatchers.filter(({ name }) =>
          routeNames.includes(name as RouteName),
        );

        return {
          getCurrentValue: () => {
            const routes = matchAll(getCurrentLocation(), matchers).reverse();
            return JSON.stringify(routes);
          },
          subscribe,
        };
      }, [JSON.stringify(routeNames)]),
    );

    return JSON.parse(routes);
  };

  const useRoute = <RouteName extends keyof FiniteRoutes | keyof NestedRoutes>(
    routeNames: ReadonlyArray<RouteName>,
  ): RouteName extends string
    ? { name: RouteName; params: Simplify<RoutesParams[RouteName]> } | undefined
    : never => {
    const route = useSubscription(
      React.useMemo(() => {
        const matchers = rankedMatchers.filter(({ name }) =>
          routeNames.includes(name as RouteName),
        );

        return {
          getCurrentValue: () => {
            const route = match(getCurrentLocation(), matchers);
            return route ? JSON.stringify(route) : route;
          },
          subscribe,
        };
      }, [JSON.stringify(routeNames)]),
    );

    return route ? JSON.parse(route) : route;
  };

  // Kudos to https://github.com/remix-run/react-router/pull/7998
  const useLink = ({
    href,
    replace = false,
    target,
  }: {
    href: string;
    replace?: boolean | undefined;
    target?: React.HTMLAttributeAnchorTarget | undefined;
  }) => {
    const active = useSubscription(
      React.useMemo(
        () => ({
          getCurrentValue: () => href === getCurrentLocation().url,
          subscribe,
        }),
        [href],
      ),
    );
    const shouldReplace = replace || active;
    const shouldIgnoreTarget = !target || target === "_self";

    return {
      active,
      onClick: React.useCallback(
        (event: React.MouseEvent) => {
          if (
            !event.defaultPrevented &&
            shouldIgnoreTarget && // Let browser handle "target=_blank" etc.
            event.button === 0 && // Ignore everything but left clicks
            !(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey) // Ignore clicks with modifier keys
          ) {
            event.preventDefault();

            if (shouldReplace) {
              unsafeReplace(href);
            } else {
              unsafeNavigate(href);
            }
          }
        },
        [shouldReplace, shouldIgnoreTarget, href],
      ),
    };
  };

  const useBlocker = (blocked: boolean, message = blockerMessage) => {
    React.useEffect(() => {
      if (!blocked) {
        return;
      }

      const unblock = history.block((transition) => {
        if (window.confirm(message)) {
          unblock();
          transition.retry();
        }
      });

      return unblock;
    }, [blocked]);
  };

  type RouteFocusProps = {
    route?:
      | { name: string; params: Record<string, string | string[] | undefined> }
      | undefined;
    containerRef: React.RefObject<unknown>;
  };

  const useRouteFocus = ({ route, containerRef }: RouteFocusProps) => {
    React.useEffect(() => {
      const element = containerRef.current as HTMLElement | undefined;
      // Only focus after a history change for UX, so that areas outside routing (e.g. navigation header)
      // are available immediately to keyboard navigation
      if (element && hasLocationChanged()) {
        try {
          const name = element.nodeName;
          // A tabIndex of -1 allows element to be programmatically focused but
          // prevents keyboard focus, so we don't want to set the value on elements
          // that support keyboard focus by default.
          if (
            element.getAttribute("tabIndex") == null &&
            focusableElements[name] == null
          ) {
            element.setAttribute("tabIndex", "-1");
          }
          element.focus();
        } catch (err) {}
      }
    }, [route?.name, JSON.stringify(route?.params), containerRef]);
  };

  return {
    getLocation: getCurrentLocation,
    createURL,
    goBack,
    goForward,
    navigate,
    replace,
    subscribe,
    unsafeNavigate,
    unsafeReplace,
    useLink,
    useLocation,
    useRoutes,
    useRoute,
    useBlocker,
    useRouteFocus,
  };
};
