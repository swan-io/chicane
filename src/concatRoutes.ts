import { parsePath } from "history";
import { RouteObject } from "./types";

const addPrefixOnNonEmpty = (value: string, prefix: string) =>
  value === "" ? value : prefix + value;

const removePrefixAndSuffix = (value: string, char: string) => {
  const unsuffixed =
    value[value.length - 1] === char ? value.slice(0, -1) : value;
  return unsuffixed[0] === char ? value.slice(1) : unsuffixed;
};

export const extractRoute = (route: string): RouteObject => {
  const { pathname: path = "", search = "", hash = "" } = parsePath(route);
  return { path, search: search.substring(1), hash: hash.substring(1) };
};

export const concatRoutes = (
  routeA: RouteObject,
  routeB: RouteObject,
): string => {
  const fixedPathA = removePrefixAndSuffix(routeA["path"], "/");
  const fixedPathB = removePrefixAndSuffix(routeB["path"], "/");
  const path = "/" + removePrefixAndSuffix(fixedPathA + "/" + fixedPathB, "/");

  const fixedSearchA = removePrefixAndSuffix(routeA["search"], "&");
  const fixedSearchB = removePrefixAndSuffix(routeB["search"], "&");
  const search = removePrefixAndSuffix(fixedSearchA + "&" + fixedSearchB, "&");

  const hash = routeB["hash"] === "" ? routeA["hash"] : routeB["hash"];

  return (
    path + addPrefixOnNonEmpty(search, "?") + addPrefixOnNonEmpty(hash, "#")
  );
};
