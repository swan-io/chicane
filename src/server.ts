import { createElement } from "react";
import { GetUniversalLocationProvider, decodeLocation } from "./history";

type Props = {
  children: React.ReactNode;
  value: string;
};

export const ServerUrlProvider = ({ children, value }: Props) => {
  const location = decodeLocation(value);

  return createElement(GetUniversalLocationProvider, {
    children,
    value: () => location,
  });
};
