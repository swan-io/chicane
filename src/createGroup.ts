import { concatRoutes } from "./concatRoutes";
import { parseRoute } from "./history";
import { ConcatRoutes } from "./types";

export const createGroup = <
  GroupName extends string,
  BaseRoute extends string,
  Routes extends Record<string, string>,
>(
  name: GroupName,
  baseRoute: BaseRoute,
  routes: Readonly<Routes>,
) => {
  const baseParsedRoute = parseRoute(baseRoute);
  const output: Record<string, string> = {};

  for (const key in routes) {
    if (Object.prototype.hasOwnProperty.call(routes, key)) {
      output[name + key] = concatRoutes(
        baseParsedRoute,
        parseRoute(routes[key]),
      );
    }
  }

  return output as {
    [K in keyof Routes as K extends string
      ? `${GroupName}${K}`
      : never]: ConcatRoutes<BaseRoute, Routes[K]>;
  };
};
