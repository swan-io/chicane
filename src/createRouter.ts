import { createPath } from "history";
import * as React from "react";
import { useSyncExternalStoreWithSelector } from "use-sync-external-store/shim/with-selector";
import { concatRoutes, extractRoute } from "./concatRoutes";
import { areRouteEqual, first, identity } from "./helpers";
import {
  history,
  subscribeToLocation,
  useGetUniversalLocation,
} from "./history";
import { getMatcher, match, matchToHistoryPath } from "./matcher";
import {
  ExtractRoutesParams,
  GetAreaRoutes,
  Matcher,
  Params,
  ParamsArg,
  PrependBasePath,
  RouteObject,
  Simplify,
} from "./types";

export const createRouter = <
  Routes extends Record<string, string>,
  BasePath extends string = string,
>(
  routes: Readonly<Routes>,
  options: {
    basePath?: BasePath;
  } = {},
) => {
  type RoutesWithBasePath = PrependBasePath<BasePath, Routes>;
  type AreaRoutes = GetAreaRoutes<RoutesWithBasePath>;
  type AreaRoutesParams = ExtractRoutesParams<AreaRoutes>;
  type FiniteRoutes = Omit<RoutesWithBasePath, keyof AreaRoutes>;
  type FiniteRoutesParams = ExtractRoutesParams<FiniteRoutes>;
  type RoutesParams = AreaRoutesParams & FiniteRoutesParams;

  const { basePath = "" } = options;

  const basePathObject: RouteObject = {
    path: extractRoute(basePath).path,
    search: "", // search and hash are not supported in basePath
    hash: "",
  };

  const matchers = {} as Record<keyof Routes, Matcher>;
  const rankedMatchers: Matcher[] = []; // higher to lower

  for (const routeName in routes) {
    if (Object.prototype.hasOwnProperty.call(routes, routeName)) {
      const matcher = getMatcher(
        routeName,
        basePath !== ""
          ? concatRoutes(basePathObject, extractRoute(routes[routeName]))
          : routes[routeName],
      );

      matchers[routeName] = matcher;
      rankedMatchers.push(matcher);
    }
  }

  rankedMatchers.sort(
    (matcherA, matcherB) => matcherB.ranking - matcherA.ranking,
  );

  const createURLFunctions = {} as {
    [RouteName in keyof FiniteRoutes]: (
      ...args: ParamsArg<FiniteRoutesParams[RouteName]>
    ) => string;
  };

  for (let index = 0; index < rankedMatchers.length; index++) {
    const matcher = rankedMatchers[index];

    if (matcher != null && !matcher.isArea) {
      const routeName = matcher.name as keyof FiniteRoutes;

      createURLFunctions[routeName] = (params?: Params) =>
        createPath(matchToHistoryPath(matchers[routeName], params));
    }
  }

  const useRoute = <RouteName extends keyof FiniteRoutes | keyof AreaRoutes>(
    routeNames: ReadonlyArray<RouteName>,
  ): RouteName extends string
    ?
        | {
            key: string;
            name: RouteName;
            params: Simplify<RoutesParams[RouteName]>;
          }
        | undefined
    : never => {
    const matchers = React.useMemo(
      () =>
        rankedMatchers.filter(({ name }) =>
          routeNames.includes(name as RouteName),
        ),
      [JSON.stringify(routeNames)],
    );

    const getUniversalLocation = useGetUniversalLocation();
    const getMatch = () => match(getUniversalLocation(), matchers);

    // @ts-expect-error
    return useSyncExternalStoreWithSelector(
      subscribeToLocation,
      getMatch,
      getMatch,
      identity,
      areRouteEqual,
    );
  };

  const push = <RouteName extends keyof FiniteRoutes>(
    routeName: RouteName,
    ...args: ParamsArg<FiniteRoutesParams[RouteName]>
  ): void => history.push(matchToHistoryPath(matchers[routeName], first(args)));

  const replace = <RouteName extends keyof FiniteRoutes>(
    routeName: RouteName,
    ...args: ParamsArg<FiniteRoutesParams[RouteName]>
  ): void =>
    history.replace(matchToHistoryPath(matchers[routeName], first(args)));

  return {
    useRoute,
    push,
    replace,
    ...createURLFunctions,
  };
};
