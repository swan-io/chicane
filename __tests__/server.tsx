/**
 * @vitest-environment node
 */

import * as React from "react";
import * as ReactDOMServer from "react-dom/server";
import { expect, test } from "vitest";
import { createRouter, Link, ServerUrlProvider } from "../src";

const Router = createRouter({
  Home: "/",
  UserList: "/users",
  UserDetail: "/users/:userId",
});

const App = () => {
  const route = Router.useRoute(["UserList", "UserDetail"]);

  return (
    <div>
      <Link to={Router.UserDetail({ userId: "123" })} activeClassName="active">
        User
      </Link>

      <div>
        {(() => {
          switch (route?.name) {
            case "UserList":
              return "UserList";
            case "UserDetail":
              return "UserDetail";
            default:
              return "404";
          }
        })()}
      </div>
    </div>
  );
};

test("Should render correctly", () => {
  expect(
    ReactDOMServer.renderToString(
      <ServerUrlProvider value="/users/123">
        <App />
      </ServerUrlProvider>,
    ),
  ).toMatchSnapshot();
});

// TODO: Add an hydration test once this library targets React 18+
