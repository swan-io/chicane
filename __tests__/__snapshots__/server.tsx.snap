// Vitest Snapshot v1

exports[`Should render home page correctly 1`] = `"<div><a href=\\"/\\" class=\\"active\\">Home</a><a href=\\"/users\\">Users</a><a href=\\"/users/123\\">User</a><div>Home</div></div>"`;

exports[`Should render user page correctly 1`] = `"<div><a href=\\"/\\">Home</a><a href=\\"/users\\">Users</a><a href=\\"/users/123\\" class=\\"active\\">User</a><div>User 123</div></div>"`;

exports[`Should render users page correctly 1`] = `"<div><a href=\\"/\\">Home</a><a href=\\"/users\\" class=\\"active\\">Users</a><a href=\\"/users/123\\">User</a><div>Users</div></div>"`;
